cmd/operator-opamp-bridge/agent/agent.go

```diff
@@ -0,0 +1,294 @@
+// store the received configuration hash regardless of application status as per the OpAMP spec.
+//
+// INVARIANT: The caller must verify that config isn't nil _and_ the configuration has changed between calls.
+func (agent *Agent) applyRemoteConfig(config *protobufs.AgentRemoteConfig) (*protobufs.RemoteConfigStatus, error) {
+	var multiErr error
+	// Apply changes from the received config map
+	for key, file := range config.Config.GetConfigMap() {
+		if len(key) == 0 || len(file.Body) == 0 {
+			continue
+		}
+		colKey, err := collectorKeyFromKey(key)
+		if err != nil {
+			multiErr = multierr.Append(multiErr, err)
+			continue
+		}
+		err = agent.applier.Apply(colKey.name, colKey.namespace, file)
+		if err != nil {
+			multiErr = multierr.Append(multiErr, err)
+			continue
+		}
+		agent.appliedKeys[colKey] = true
+	}
+	// Check if anything was deleted
+	for collectorKey := range agent.appliedKeys {
+		if _, ok := config.Config.GetConfigMap()[collectorKey.String()]; !ok {
+			err := agent.applier.Delete(collectorKey.name, collectorKey.namespace)
+			if err != nil {
+				multiErr = multierr.Append(multiErr, err)
+			}
+		}
+	}
+	agent.lastHash = config.GetConfigHash()
+	if multiErr != nil {
+		return &protobufs.RemoteConfigStatus{
+			LastRemoteConfigHash: agent.lastHash,
+			Status:               protobufs.RemoteConfigStatuses_RemoteConfigStatuses_FAILED,
+			ErrorMessage:         multiErr.Error(),
+		}, multiErr
+	}
+	return &protobufs.RemoteConfigStatus{
+		LastRemoteConfigHash: agent.lastHash,
+		Status:               protobufs.RemoteConfigStatuses_RemoteConfigStatuses_APPLIED,
+	}, nil
+}
+
+// Shutdown will stop the OpAMP client gracefully.
+func (agent *Agent) Shutdown() {
+	agent.logger.Debugf("Agent shutting down...")
+	if agent.opampClient != nil {
+		err := agent.opampClient.Stop(context.Background())
+		if err != nil {
+			agent.logger.Errorf(err.Error())
+		}
+	}
+	if agent.metricReporter != nil {
+		agent.metricReporter.Shutdown()
+	}
+}
+
+// onMessage is called when the client receives a new message from the connected OpAMP server. The agent is responsible
+// for checking if it should apply a new remote configuration. The agent will also initialize metrics based on the
+// settings received from the server. The agent is also able to update its identifier if it needs to.
+func (agent *Agent) onMessage(ctx context.Context, msg *types.MessageData) {
+
+	// If we received remote configuration, and it's not the same as the previously applied one
+	if msg.RemoteConfig != nil && !bytes.Equal(agent.lastHash, msg.RemoteConfig.GetConfigHash()) {
+		var err error
+		status, err := agent.applyRemoteConfig(msg.RemoteConfig)
+		if err != nil {
+			agent.logger.Errorf(err.Error())
+		}
+		err = agent.opampClient.SetRemoteConfigStatus(status)
+		if err != nil {
+			agent.logger.Errorf(err.Error())
+			return
+		}
+		err = agent.opampClient.UpdateEffectiveConfig(ctx)
```

Comment:
Note that this will error if `ReportsEffectiveConfig` capability is not specified. It may result in unnecessary error message in the log, so perhaps check if the user wants the capability and only in that case call this function.

Reply:
I see this is guarded by `remoteConfigEnabled` which is populated from `AcceptsRemoteConfig` capability. However `AcceptsRemoteConfig` and `ReportsEffectiveConfig` are different capabilities and can be set independently. It is possible for ReportsEffectiveConfig=true and ReportsEffectiveConfig=false (although it is a great config for users to use and I don't know why anyone would want to use this combination of capabilities).